### ISSUE TITLE
XEP-0198 implementation optimizations and errors

### ISSUE THREAD

**Issue by tmolitor-stud-tu** [anurodhp/Monal#264]
I observed Monal is requesting acks quite often (_requesting them even if the counter of outgoing stanzas did not change_).
Ideally a smacks ack request should only be sent if no other ack is pending and the outgoing queue has still items.
Nonzas should **not** trigger an ack request!

**Additionally the following implementation error amounts for many unneeded XEP-0357 push requests because the outgoing queue on the server might not be fully acked despite all stanzas were already received by Monal:**

XEP-0198 says in its revision history (https://xmpp.org/extensions/xep-0198.html#revision-history-v1.5.2):
> Send 'a' element before stream closure.

Which maps to this part in section 4:
>  It is RECOMMENDED that initiating entities (usually clients) send an <a/> element right before they gracefully close the stream, in order to inform the peer about received stanzas. Otherwise it can happen that stanzas are re-sent (usually by the server) although they were actually received.

**Comment by tmolitor-stud-tu** [comment]
Example for too much ack requests:

```
Jul 17 16:06:03 c2s558a42a5e5d0 debug   SEND(33): <a xmlns='urn:xmpp:sm:3' h='19'/>
[... one new message stanza incoming ...]
Jul 17 16:06:03 c2s558a42a5e5d0 debug   SEND(26): <r xmlns='urn:xmpp:sm:3'/>
Jul 17 16:06:03 c2s558a42a5e5d0 debug   RECV(30): <r xmlns='urn:xmpp:sm:3' ></r>
Jul 17 16:06:03 c2s558a42a5e5d0 debug   SEND(33): <a xmlns='urn:xmpp:sm:3' h='20'/>
Jul 17 16:06:03 c2s558a42a5e5d0 debug   RECV(60): <r xmlns='urn:xmpp:sm:3' ></r><r xmlns='urn:xmpp:sm:3' ></r>
```

Clearly only one of the 3(!!) requests is needed here.

**Comment by tmolitor-stud-tu** [comment]
Complete log after resume (grepped to contain only smacks `nonzas`):

```
Jul 17 16:06:03 c2s558a42a5e5d0 debug   SEND(85): <resumed h='14' xmlns='urn:xmpp:sm:3' previd='942bde71-ad88-4046-b3c8-49c44ffc9762'/>
Jul 17 16:06:03 c2s558a42a5e5d0 debug   SEND(26): <r xmlns='urn:xmpp:sm:3'/>
Jul 17 16:06:03 c2s558a42a5e5d0 debug   RECV(1278): <a xmlns='urn:xmpp:sm:3'  h='29' ></a><inactive xmlns='urn:xmpp:csi:0' ></inactive><r xmlns='urn:xmpp:sm:3' ></r><r xmlns='urn:xmpp:sm:3' ></r><r xmlns='urn:xmpp:sm:3' ></r>
Jul 17 16:06:03 c2s558a42a5e5d0 debug   SEND(33): <a xmlns='urn:xmpp:sm:3' h='19'/>
Jul 17 16:06:03 c2s558a42a5e5d0 debug   SEND(33): <a xmlns='urn:xmpp:sm:3' h='19'/>
Jul 17 16:06:03 c2s558a42a5e5d0 debug   SEND(33): <a xmlns='urn:xmpp:sm:3' h='19'/>
Jul 17 16:06:03 c2s558a42a5e5d0 debug   SEND(26): <r xmlns='urn:xmpp:sm:3'/>
Jul 17 16:06:03 c2s558a42a5e5d0 debug   RECV(30): <r xmlns='urn:xmpp:sm:3' ></r>
Jul 17 16:06:03 c2s558a42a5e5d0 debug   SEND(33): <a xmlns='urn:xmpp:sm:3' h='20'/>
Jul 17 16:06:03 c2s558a42a5e5d0 debug   RECV(60): <r xmlns='urn:xmpp:sm:3' ></r><r xmlns='urn:xmpp:sm:3' ></r>
Jul 17 16:06:03 c2s558a42a5e5d0 debug   SEND(33): <a xmlns='urn:xmpp:sm:3' h='20'/>
Jul 17 16:06:03 c2s558a42a5e5d0 debug   SEND(33): <a xmlns='urn:xmpp:sm:3' h='20'/>
Jul 17 16:06:03 c2s558a42a5e5d0 debug   RECV(38): <a xmlns='urn:xmpp:sm:3'  h='29' ></a>
Jul 17 16:06:03 c2s558a42a5e5d0 debug   RECV(38): <a xmlns='urn:xmpp:sm:3'  h='32' ></a>
[the next one is okay and in fact wanted (acking all stanzas before closing the connection)]
Jul 17 16:06:23 c2s558a42a5e5d0 debug   RECV(38): <a xmlns='urn:xmpp:sm:3'  h='32' ></a>
Jul 17 16:06:23 c2s558a42a5e5d0 debug   RECV(45): <inactive xmlns='urn:xmpp:csi:0' ></inactive>
Jul 17 16:06:30 c2s558a42a5e5d0 info    Client disconnected: closed
```

**Comment by tmolitor-stud-tu** [comment]
still not fixed in monal 3.8.1
